### PR TITLE
Refactored checks in data structures

### PR DESCRIPTION
**Pull request by iulianpascalau**
- added extra checks (moved from mx-chain-go in mx-chain-core-go)

### CODE REVIEW

File: data/block/block.go [hunk 1]
```diff
@@ -581,3 +582,39 @@ func (h *Header) GetAdditionalData() headerVersionData.HeaderAdditionalData {
 	// no extra data for the initial version of shard block header
 	return nil
 }
+
+// CheckFieldsForNil checks a predefined set of fields for nil values
+func (h *Header) CheckFieldsForNil() error {
```
Comment (iulianpascalau) [inline data/block/block.go:587]: missing addition on the `HeaderHandler` interface

File: data/block/block.go [hunk 2]
```diff
@@ -581,3 +582,39 @@ func (h *Header) GetAdditionalData() headerVersionData.HeaderAdditionalData {
...
+	if h == nil {
+		return data.ErrNilPointerReceiver
+	}
+	if h.PrevHash == nil {
```
Comment (SebastianMarian) [inline data/block/block.go:591]: PrevHash, PrevRandSeed and others fields checked here of the genesis block are not nil?

File: data/block/blockV2.go [hunk 3]
```diff
@@ -629,3 +630,25 @@ func (hv2 *HeaderV2) GetAdditionalData() headerVersionData.HeaderAdditionalData
...
+// CheckFieldsForNil checks a predefined set of fields for nil values
+func (hv2 *HeaderV2) CheckFieldsForNil() error {
+	if hv2 == nil {
+		return data.ErrNilPointerReceiver
+	}
+	err := hv2.Header.CheckFieldsForNil()
+	if err != nil {
+		return err
+	}
+	if hv2.ScheduledRootHash == nil {
```
Comment (SebastianMarian) [inline data/block/blockV2.go:643]: All the fields checked here of the genesis block are not nil?

File: data/block/metaBlock.go [hunk 4]
```diff
@@ -523,3 +524,45 @@ func (m *MetaBlock) GetAdditionalData() headerVersionData.HeaderAdditionalData {
 	// no extra data for the initial version of meta block header
 	return nil
 }
+
+// CheckFieldsForNil checks a predefined set of fields for nil values
+func (m *MetaBlock) CheckFieldsForNil() error {
+	if m == nil {
+		return data.ErrNilPointerReceiver
+	}
+	if m.PrevHash == nil {
```
Comment (SebastianMarian) [inline data/block/metaBlock.go:533]: PrevHash, PrevRandSeed and others fields checked here of the genesis block are not nil?

File: data/block/block.go (hunk 1 above)
Comment (raduchis) [inline data/block/block.go:587]: are all these checks backwards compatible? did we have some of them nil sometimes and now they cannot be nil?

File: data/block/block.go [hunk 5]
```diff
@@ -581,3 +582,39 @@ func (h *Header) GetAdditionalData() headerVersionData.HeaderAdditionalData {
...
+	}
+	if h.SoftwareVersion == nil {
+		return fmt.Errorf("%w in Header.SoftwareVersion", data.ErrNilValue)
+	}
+	if h.AccumulatedFees == nil {
+		return fmt.Errorf("%w in Header.AccumulatedFees", data.ErrNilValue)
+	}
+	if h.DeveloperFees == nil {
+		return fmt.Errorf("%w in Header.DeveloperFees", data.ErrNilValue)
+	}
```
Comment (raduchis) [inline data/block/block.go:617]: Signature and LeaderSignature should not be checked for nil? both here and in the metablock?

**Review by raduchis — COMMENTED** [review]
just small clarifications

File: data/block/block.go (hunk 2 above)
Comment (iulianpascalau) [reply data/block/block.go:591]: as seen here, no, they are populated: https://github.com/multiversx/mx-chain-go/blob/10c4bef8b57431eee74fa0d34af29c86a05f58c6/genesis/process/shardGenesisBlockCreator.go#L207
https://github.com/multiversx/mx-chain-go/blob/10c4bef8b57431eee74fa0d34af29c86a05f58c6/genesis/process/metaGenesisBlockCreator.go#L97

File: data/block/blockV2.go (hunk 3 above)
Comment (iulianpascalau) [reply data/block/blockV2.go:643]: we do not have genesis block on the version 2 of the block

File: data/block/metaBlock.go (hunk 4 above)
Comment (iulianpascalau) [reply data/block/metaBlock.go:533]: they are not nil

File: data/block/block.go (hunk 1 above)
Comment (iulianpascalau) [reply data/block/block.go:587]: will definitely check with an import-db. At the first sight, they should be backwards compatible

File: data/block/block.go (hunk 5 above)
Comment (iulianpascalau) [reply data/block/block.go:617]: I do not think so. Reason: the proposed block that is sent on the consensus topic does not set those 2 fields

**Review by raduchis — DISMISSED** [review]

File: data/block/block.go (hunk 1 above)
Comment (raduchis) [reply data/block/block.go:587]: ok

File: data/block/block.go (hunk 5 above)
Comment (raduchis) [reply data/block/block.go:617]: ok

File: data/block/block.go (hunk 2 above)
Comment (SebastianMarian) [reply data/block/block.go:591]: ok

File: data/block/metaBlock.go (hunk 4 above)
Comment (SebastianMarian) [reply data/block/metaBlock.go:533]: ok

File: data/block/blockV2.go (hunk 3 above)
Comment (SebastianMarian) [reply data/block/blockV2.go:643]: Yep, ok

File: data/block/block.go (hunk 5 above)
Comment (SebastianMarian) [reply data/block/block.go:617]: Also PubKeyBitmap is nil on the proposed header.

File: data/block/block.go [hunk 6]
```diff
@@ -581,3 +582,39 @@ func (h *Header) GetAdditionalData() headerVersionData.HeaderAdditionalData {
...
+	if h.PrevHash == nil {
+		return fmt.Errorf("%w in Header.PrevHash", data.ErrNilValue)
+	}
+	if h.PrevRandSeed == nil {
+		return fmt.Errorf("%w in Header.PrevRandSeed", data.ErrNilValue)
+	}
+	if h.RandSeed == nil {
+		return fmt.Errorf("%w in Header.RandSeed", data.ErrNilValue)
+	}
+	if h.PubKeysBitmap == nil {
```
Comment (SebastianMarian) [inline data/block/block.go:600]: This field is nil for the proposed headers!

File: data/block/block.go (hunk 5 above)
Comment (sasurobert) [reply data/block/block.go:617]: +1

File: data/block/block.go (hunk 6 above)
Comment (iulianpascalau) [reply data/block/block.go:600]: Ok, will fix

File: data/block/blockV2.go (hunk 3 above)
Comment (sasurobert) [inline data/block/blockV2.go:643]: Can Scheduled roothash be nil? Or it is equal with roothash? On sovereign this is certainly nil

Comment (iulianpascalau) [reply data/block/blockV2.go:643]: will remove this check, it is handled in the scheduled processor

File: data/block/blockV2.go [hunk 7]
```diff
@@ -640,9 +640,7 @@ func (hv2 *HeaderV2) CheckFieldsForNil() error {
 	if err != nil {
 		return err
 	}
-	if hv2.ScheduledRootHash == nil {
-		return fmt.Errorf("%w in HeaderV2.ScheduledRootHash", data.ErrNilValue)
```
Comment (SebastianMarian) [inline data/block/blockV2.go]: Is there a situation in HeaderV2 where ScheduledRootHash is nil? It is set on CreateNewHeader, so also on the proposed headers, with the value of the current rootHash.

**Review by SebastianMarian — APPROVED** [review]

**Review by raduchis — APPROVED** [review]

File: data/block/blockV2.go (hunk 7 above)
Comment (iulianpascalau) [reply data/block/blockV2.go]: Was removed for sovereign compatibility, as @sasurobert mentioned

File: data/block/block.go [hunk 8]
```diff
@@ -581,3 +582,36 @@ func (h *Header) GetAdditionalData() headerVersionData.HeaderAdditionalData {
...
+	if h.RandSeed == nil {
+		return fmt.Errorf("%w in Header.RandSeed", data.ErrNilValue)
+	}
+	if h.RootHash == nil {
+		return fmt.Errorf("%w in Header.RootHash", data.ErrNilValue)
+	}
+	if h.ChainID == nil {
+		return fmt.Errorf("%w in Header.ChainID", data.ErrNilValue)
+	}
+	if h.SoftwareVersion == nil {
```
Comment (AdoAdoAdo) [inline data/block/block.go:606]: this one don't remember if it was populated from the start, (genesis?)

**Review by AdoAdoAdo — APPROVED** [review]

File: data/block/block.go (hunk 8 above)
Comment (iulianpascalau) [reply data/block/block.go:606]: I will check and come back with an answer to this. As far as I remember, we've populated this from the beginning.

Comment (iulianpascalau) [reply data/block/block.go:606]: It is populated with an empty slice. As far as I've tested, the genesis + a few hundred blocks afterward, and everything went ok. Started the import-db process nevertheless.

**Pull request merged**